File: Back/NodeJS/modules/bets/src/infrastructure/repositories/sql/SqlMemberRepository.ts
import { DomainEventAccessor } from "../../../../../shared/infrastructure/events/DomainEventAccessor";
import { IMemberRepository } from "../../../domain/members/IMemberRepository";
import { Member } from "../../../domain/members/Member";
import { MemberId } from "../../../domain/members/MemberId";
import { MemberSnapshot } from "../../../domain/members/MemberSnapshot";
import { BetContext } from "../../DataAccess/BetContext";
import { MemberEntity } from "../../DataAccess/entities/MemberEntity";

export class SqlMemberRepository implements IMemberRepository {

    constructor(private readonly betContext: BetContext, 
                        private readonly domainEventAccessor: DomainEventAccessor
            ){}

            
    async Save(member: Member): Promise<void> {
        const entity = MemberEntity.Create(member.GetSnapshot());
        await this.betContext.MemberEntity.save(entity);
        this.domainEventAccessor.AddEvents(member.DomainEvents);
    }
    async GetByIdAsync(memberId: MemberId): Promise<Member | undefined> {
        const entity = await this.betContext.MemberEntity.findOneBy({
            Id: memberId.Value
        })
        if(!entity) {
            return undefined;
        }
        return Member.FromSnapshot(new MemberSnapshot(entity.Id, entity.Username, entity.Wallet, entity.CountFriends))
    }

}

File: Back/.Net/Modules/Bets/BetFriends.Bets.Infrastructure/Repositories/Sql/SqlMemberRepository.cs
using BetFriends.Bets.Domain.Members;
using BetFriends.Bets.Infrastructure.Repositories.Sql.DataAccess;
using BetFriends.Shared.Infrastructure.Event;
using Microsoft.EntityFrameworkCore;

namespace BetFriends.Bets.Infrastructure.Repositories.Sql;

internal class SqlMemberRepository(BetContext betContext, DomainEventsAccessor domainEventsAccessor) : IMemberRepository
{
    private readonly BetContext betContext = betContext;
    private readonly DomainEventsAccessor domainEventsAccessor = domainEventsAccessor;

    public async Task<Member> GetByIdAsync(MemberId memberId)
    {
        var entity = await betContext.Members.FindAsync(memberId.Value);
        if (entity != null)
            return Member.FromState(new MemberState(entity.MemberId,
                                                    entity.Username,
                                                    entity.Wallet,
                                                    entity.CountFriends));
        return default!;
    }

    public async Task<IEnumerable<Member>> GetByIdsAsync(IEnumerable<Guid> guests)
    {
        var members = await betContext.Members.Where(x => guests.Contains(x.MemberId)).ToListAsync();
        return members.Select(x => Member.FromState(new MemberState(x.MemberId,
                                                                    x.Username,
                                                                    x.Wallet,
                                                                    x.CountFriends)));
    }

    public async Task<bool> IsMemberExistAsync(MemberId memberId)
    {
        return (await betContext.Members.CountAsync(x => x.MemberId == memberId.Value)) != 0;
    }

    public async Task SaveAsync(Member member)
    {
        var entity = await betContext.Members.FindAsync(member.MemberId.Value);
        if (entity == null)
        {
            await betContext.Members.AddAsync(new MemberEntity(member));
            return;
        }
        entity.Update(member.Snapshot);
        domainEventsAccessor.Add(member.Events);
    }

    public async Task SaveAsync(IEnumerable<Member> gamblers)
    {
        var entities = await betContext.Members
                                     .Where(x => gamblers.Any(g => g.Snapshot.MemberId == x.MemberId))
                                     .ToListAsync();
        foreach (var entity in entities)
        {
            var gambler = gamblers.FirstOrDefault(x => x.Snapshot.MemberId == entity.MemberId);
            if (gambler != null)
                entity.Update(gambler.Snapshot);
        }
    }
}
